Levelup

balazs482 opened the pull request:
@oliverosz
Készen van az új osztály és a tesztek.

A main-ben az eddigi két Character helyett egy Character és egy Player objektumot ütköztetek, hogy használat közben is látszódjon az új osztály. Ennek megfelelően felülírtam az expected_output.csv fájlt (amit a diff-hez használunk).

Eredeti tervem szerint a Character osztály DMG tagváltozója const float típusú lett volna, amit aztán non-constként újradeklarálok a Playerben. Bármiféle javaslatot előre köszönök azt illetően, hogy ez hogy nem dobna warningot.  (Most sehol nem const, és csak egyszer van létrehozva.)

A main.cpp valamiért nem akar jól indentálódni. GitHubon próbáltam módosítani tabulátoros és szóközös módban is, de mindig elugrott valami új helyen. Most a 8 sor. Illetve én 4 szóköznyi méretet használtem, de többnyire 8 lett (kivéve valamiért a 24-es és 28-as sorokat). Erre esetleg valami praktika? Ilyeneket tényleg nem szeretnék benthagyni.

balazs482 committed:
Intigers coverted to floats

balazs482 committed:
maxHP is new attribute of Character Class

Maximum hitpoint value is stored separate from current HP. All Characters are constructed with having their maximum HP.

balazs482 committed:
Character class access modifiers changed

maxHP, HP and DMG are now protected, because planned subclass function levelup() changes them.

balazs482 committed:
Introducing Player class

Player class is a subclass of Character. It has XP and can level up.

balazs482 committed:
XP gaining implemented

levelup() function now in private, gainXP() uses it.

balazs482 committed:
Math.h inclusion

round() is used, but the header had been forgotten about.

balazs482 committed:
Merge branch 'master' into levelup

balazs482 committed:
Numeric type differences fixed

Used master's files while update merging, which did not use floats in Character class, but its Player subclass did.

balazs482 committed:
gotHit() overloading

Now if the parameter (attacker) of gotHit() is a player object, the gainXP() is called on it. With future subclasses of Character the assertion may require changes.

balazs482 committed:
Heritage type of Player class changed

balazs482 committed:
Critical member function fixes

Player constructor rewritten, inclusion modified

balazs482 committed:
Access specifier changed

Now everything is available with subclass.

balazs482 committed:
Value assigning method changed

Shorter form is used for an easier read.

balazs482 committed:
Merge branch 'levelup' of https://github.com/Teaching-projects/SZE-MOSZE-2020-Mortar_Kombat into levelup

balazs482 committed:
Type-o changed

Compound assignment in gotHit() suffered from a typing mistake previously.

balazs482 committed:
Hit algorythm and class construction reworked,

gotHit() is now used in doHit(), which is the public function for attacks. The only public construction method is now via [classname]FromFile() which takes the arguments right away in main().

balazs482 committed:
Conflict solved

gotHit() return had a conflict

balazs482 committed:
main() now uses Player class

One of the Character objects in main() is changed to Player type.

balazs482 committed:
Updated the expected output (levelup stats)

Updated expected_output.csv, because one of the characters now belongs to Player subclass thus level ups (and obviously wins with current starting values).

balazs482 committed:
Character and Player classes updated

balazs482 committed:
Create code_analysis.sh

balazs482 committed:
Test updated, now has Ccpcheck

Test now terminates on errors and warnings. Test artifacts style and performance notes.

balazs482 committed:
Update expected_output.csv

balazs482 committed:
Create memtest.sh

balazs482 committed:
Test now has Valgrind memtest

balazs482 committed:
Indentation and blank line fix

balazs482 committed:
Indentation fix

oliverosz commented on Player.cpp:
```diff
@@ -0,0 +1,33 @@
+
+void Player::doHit(Character& victim)
+{
+	this->gainXP(victim.gotHit(this));
```
Nem kéne ezt függővé tenni a `Character::gotHit` visszatérési értékétől. Inkább külön legyen meghívva a `gainXP` a damage-gel.

balazs482 replied:
Azért is döntöttem így, hogy ne kelljen kétszer kiszámolni azt, hogy a támadó teljes sebzése, vagy az áldozat maradék hp-ja a kapott XP. Na meg így egy függvényt kell hívni egy támadásra.   
Vagy csak az úgy legyen, hogy a doHit() függvényen belül ne egyenesen a paraméter legyen a gotHit(), hanem előbb egy <float> experiencePoints (vagy inflictedDamage) változót hozzak létre neki, és aztán
```
this->gainXP(experiencePoints)
```
?

balazs482 replied:
**A**   
A gotHit() public marad visszatérési érték nélkül. A gainXP()-t virtuálisként létre kell hozni a Character osztályban, hogy meg lehessen arra is hívni, csak akkor ne utaljon a csak az alosztályban létező XP tagváltozóra. Ebben az esetben a doHit() feleslegessé válik. A Player::gainXP() overloadban ki kellene számolni újra a bevitt HP-t, mint potenciális XP-t a növelés előtt. Ehhez ugye nem float bemenet kell, hanem az áldozatot kapja majd referenciaként a meghíváskor. Aztán a main()-ben a doHit() helyett egy gainXP(), majd egy gotHit() kerül meghívásra mindig a másikra referálva.
     
**B**  
```
void Player::doHit(Character& victim)
{
        float inflictedDamage = victim.gotHit(this);
	this->gainXP(inflictedDamage);
```   
    
**C**  
Valami, amire nem gondolok egyáltalán.

balazs482 replied:
Esetleg még annyit lehet tenni, hogy a tényleges sebzés kiszámolását kiszervezem egy Character osztály beli protected  függvénybe. Ezt a támadóra kell meghívni, mert a gotDMG() protected, de a gotHP() nem (fight során csak a HP-t kell kiíratni). Így legalább a kód nem lesz redundáns, de a számítást ugyanúgy kétszer végzi el a fight minden turnusában.

oliverosz commented on Character.h:
```diff
@@ -1,17 +1,23 @@
-	const unsigned int getDMG() const;
-	void gotHit(const Character &attacker);
+	const float getHP() const;
+	float gotHit(Character* attacker);
+	virtual void doHit(Character& victim);
```
Nem jó, hogy mindkét metódus public, így nem egyértelmű az osztály használójának, hogy melyiket (esetleg mindkettőt) kéne használnia.

balazs482 replied:
Igen, ezt elfelejtettem áttenni. Javítva.

balazs482 replied:
Oh, már emlékszem miért volt public. Hiszen a victimre hívja meg a doHit(). Azért van ez a gotHit() - doHit() páros, mert az a gotHit a sebzettet módosítja, a doHit() a támadót módosít(hat)ja. Hogyan lehet protected a gotHit(), de mégis megoldva a sebzés a doHit()-en belül?

balazs482 replied:
Egyébként az a veszély, hogy valaki mindkettőt használná, tényleg fennáll. A másik commentben említett egymástól független sebző és XP kapó függvények megoldanák ezt a problémát, csak szerintem sok más szempontból bonyolultabb lenne a szerkezet. Például az is könnyen okozna akkor félreértést, hogy miért van gainXP() normál karakterre, akinek még csak nincs is XP-je. (Persze mert a fight alatt nem tudjuk, hogy ki Player és ki nem.)
   
Másfelől a külön használható gotHit() kimondottan hasznosnak tűnik.   
    
Nem jó úgy, ha írok kommenteket a Character.h-ba? Vagy később nem terjed ki úgyis erre is a Doxygen?

oliverosz replied:
>Nem jó úgy, ha írok kommenteket a Character.h-ba? Vagy később nem terjed ki úgyis erre is a Doxygen?

A Doxygen nem erre való. Olyan metódusnak nem szabadna publikusnak lennie, amit nem szabad meghívni kívülről.

Viszont amire gondoltam, az valóban nem működik itt, mert különböző osztályokról van szó. (Nem volt előírva, hogy újat kell származtatni, elég lett volna kiegészíteni a Charactert.) Mivel úgyis át kell adni egy támadót a `gotHit`nek, így nem olyan vészes, maradhat. Másik lehetőség a setter (`setHP`) használata lenne, de akkor meg azt lehetne bármivel meghívni. Elfogadom így.

balazs482 replied:
Igazából pont ma akartam megkérdezni a társam, hogy mit csinál a Doxygen, amin dolgozik. Akkor helyette hozzáadok pár sor kommentet a headerbe.

balazs482 committed:
Last attempt on indentation fix

oliverosz commented on Player.cpp:
```diff
@@ -0,0 +1,33 @@
+	while (levelupXP >= 100)
+	{
+		levelupXP -= 100;
+		HP = maxHP = static_cast<float>(round(maxHP * 1.1));
+		DMG = static_cast<float>(DMG * 1.1);
```
Nem látom sok értelmét `double`-ként kiszámolni az eredményt, ha utána úgyis át lesz castolva `float`ra. `1.1f`-et kell írni, és akkor `float`ként kezeli a lebegőpontos literált. És a DMG-nél miért nincs kerekítés?

balazs482 replied:
1. Ez az 'f' dolog nekem új, de köszönöm, javítva.

2. Csak a maxHP-nál volt kérve a kerekítés a feladatkiírásban

balazs482 committed:
Perfect indentation

I have screenshot lol

balazs482 committed:
Update README.md

Readme explains new functionality

oliverosz reviewed (commented):
>Eredeti tervem szerint a Character osztály DMG tagváltozója const float típusú lett volna, amit aztán non-constként újradeklarálok a Playerben. Bármiféle javaslatot előre köszönök azt illetően, hogy ez hogy nem dobna warningot. (Most sehol nem const, és csak egyszer van létrehozva.)

Nem lehet "újradeklarálni" tagváltozót. Ha a Character olyan osztály, amiben a DMG konstans, akkor a Playerre nem igaz, hogy az egy Character (mivel a Playernek nem konstans a sebzése), tehát nem származhat belőle.

Nem kellett volna külön Character objektumot létrehozni, de mindegy, a main úgyis fog még változni.

Amúgy rendben van, ezt a pár észrevételt kell javítani.

balazs482 committed:
gotHit() now protected

doHit() is now the only public function for attacks.

balazs482 committed:
gotHit() is back to being public again

doHit() uses gotHit() on its argument, so gotHit() must be public with the current structure.

balazs482 committed:
levelup() now uses floats in its expressions

...so there is no need for manual casting.

balazs482 commented:
@oliverosz megcsináltam, amit tudtam vagy egyértelmű volt.
A nyitó kommentre:
1. Akkor rendben van, hogy a DMG a sima karakterek esetében nem konstans?
2. Majd figyelünk a merge-eknél, hogy az én main.cpp-mből semmi ne kerüljön bele az inclusiont leszámítva, valamint a diff-hez használt fájl se legyen felülírva.

oliverosz commented:
@hegyhati kérek jogokat a merge-hez

balazs482 committed:
Comments added to clarify member functions

hegyhati commented:
@oliverosz done.
